command/agent/agent.go

```diff
@@ -593,7 +594,46 @@ func (a *Agent) setupClient() error {
 	return nil
 }
 
-// setupNodeID will pull the persisted node ID, if any, or create a random one
+// makeNodeID will try to find a host-specific ID, or else will generate a
+// random ID. The returned ID will always be formatted as a GUID. We don't
+// tell the caller whether this ID is random or stable since the consequences
+// are high for us if this changes, so we will persist it either way. This will
+// let gopsutil change implementations without affecting in-place upgrades of
+// nodes.
+func (a *Agent) makeNodeID() (string, error) {
+	var id string
+
+	// Try to get a stable ID associated with the host itself.
+	info, err := host.Info()
+	if err != nil {
+		a.logger.Printf("[DEBUG] Failed to get a unique id from the host, generating a random one: %v", err)
+		goto RANDOM_ID
+	}
+
+	// Make sure the host ID parses as a UUID, since we don't have complete
+	// control over this process.
+	id = strings.ToLower(info.HostID)
+	if _, err := uuid.ParseUUID(id); err != nil {
+		a.logger.Printf("[DEBUG] Host ID %q isn't formatted as a UUID, generating a random one: %v",
+			id, err)
+		goto RANDOM_ID
+	}
+
+	a.logger.Printf("[DEBUG] Using ID %q from host as node ID", id)
+	return id, nil
+
+	// Worst case, just generate a random UUID.
+RANDOM_ID:
```

Comment:
Looks good.  The only feedback I have is a completely optional style note.  Instead of having a `goto`, I'd use a second `func` encapsulates all of the execution of `RANDOM_ID` and have the new function return a random Node ID.

```
goto RANDOM_ID
// becomes
return makeRandomHostID()
```

If you wanted to "clever", you could:

```
var id string
defer func() { a.logger.Printf("[DEBUG] Using random UUID %q as node ID", id) }
```

and skip the manual placement of the debug message.